File: Javascript/InsertionSort.js
// Insertion Sort

var a = [4,3,7,5,10,65,1,9,2];

var b = ["c", "z", "aa", "b", "g", "a"];

console.log(a);

function InsertionSort(arr) {
	
	for (i = 0; i < arr.length; i++) {

		var j = i;
		while (j >= 0) {
			if (arr[j+1] < arr[j]) {
			var temp = arr[j];
			arr[j] = arr[j+1];
			arr[j+1] = temp;
			}
		j = j-1;
		}
	}
}

InsertionSort(a);
console.log(a)

function exch(a, v1, v2) {
	var swap = a[v1];
	a[v1] = a[v2];
	a[v2] = swap;
}

function InsertionSort2(arr) {
	var a = arr.length;

	for (i = 0; i < a; i++) {
		for (j = i; j >= 0; j--) {
			if (arr[j] > arr[j+1])
				exch(arr, j, j+1);
		}
	}
}

InsertionSort2(a);

console.log(a);
console.log(b);

InsertionSort2(b);

console.log(b);

File: ElemSorts/InsertionSort.java
public class InsertionSort {

	public static void sort(Comparable[] a) {

		int N = a.length;

		int j;

		for (i = 0; i < N; i++) {

			for (j = i; j >= 0; j--) {
				if (less(a[j], a[j-1])) {
					exch(a, j, j-1);
				}
			}
		}
	}

	private static void exch(Comparable[] a, int i, int j) {
		
		Comparable swap = a[i];
		a[i] = a[j];
		a[j] = swap;

	}

	private static boolean less(Comparable v, Comparable w) {
		return v.compareTo(w) < 0;
	}

	private static boolean isSorted(Comprable[] a) {
		for (int i = 0; i < a.length; i++)
			if (less(a[i], a[i-1]))
				return false;
			return true;
	}
}